src/OrleansRuntime/Catalog/Catalog.cs

```diff
@@ -567,130 +514,176 @@ private void SetupActivationInstance(ActivationData result, string grainType, st
+        private enum ActivationInitializationStage
+        {
+            None,
+            Register,
+            SetupState,
+            InvokeActivate,
+            Completed
+        }
+
+        private async Task InitActivation(ActivationData activation, string grainType, string genericArguments,
+            Dictionary<string, object> requestContextData)
         {
             // We've created a dummy activation, which we'll eventually return, but in the meantime we'll queue up (or perform promptly)
             // the operations required to turn the "dummy" activation into a real activation
-            ActivationAddress address = activation.Address;
+            var initStage = ActivationInitializationStage.None;
+            Exception exception = null;
+            var registrationResult = default(ActivationRegistrationResult);
 
-            int initStage = 0;
             // A chain of promises that will have to complete in order to complete the activation
             // Register with the grain directory, register with the store if necessary and call the Activate method on the new activation.
             try
             {
-                initStage = 1;
-                await RegisterActivationInGrainDirectoryAndValidate(activation);
+                initStage = ActivationInitializationStage.Register;
+                registrationResult = await RegisterActivationInGrainDirectoryAndValidate(activation);
 
-                initStage = 2;
-                await SetupActivationState(activation, String.IsNullOrEmpty(genericArguments) ? grainType : $"{grainType}[{genericArguments}]");
+                // If registration failed, recover from within the finally block.
+                if (!registrationResult.IsSuccess) return;
 
-                initStage = 3;
+                initStage = ActivationInitializationStage.SetupState;
+                await SetupActivationState(activation,
+                    string.IsNullOrEmpty(genericArguments) ? grainType : $"{grainType}[{genericArguments}]");
+
+                initStage = ActivationInitializationStage.InvokeActivate;
                 await InvokeActivate(activation, requestContextData);
 
                 ActivationCollector.ScheduleCollection(activation);
 
                 // Success!! Log the result, and start processing messages
-                if (logger.IsVerbose) logger.Verbose("InitActivation is done: {0}", address);
+                initStage = ActivationInitializationStage.Completed;
+                if (logger.IsVerbose) logger.Verbose("InitActivation is done: {0}", activation.Address);
             }
             catch (Exception ex)
             {
-                lock (activation)
+                // Store and re-throw the exception.
+                exception = ex;
+                throw;
+            }
+            finally
```

Comment:
why is this in the finally block? Why not both in the catch block as well as in the `if (!registrationResult.IsSuccess)` part?